### PR TITLE
Online video support for VLMs

### LINKED ISSUE #9842: [Feature]: Online video support for VLMs
### Online video support for VLMs
vLLM already supports a large number of MultiModal Machine Learning visual models, some of which support image and video input，such as [Qwen2-VL](https://huggingface.co/Qwen/Qwen2-VL-2B-Instruct), [LLaVA-Onevision](https://huggingface.co/llava-hf/llava-onevision-qwen2-7b-ov-hf), etc. Referring to the implementation of image, this proposal adds support for video.

Refer to the visual interfaces of OpenAI [(vision](https://platform.openai.com/docs/guides/vision)and [video](https://cookbook.openai.com/examples/gpt_with_vision_for_video_understanding)) and Google Gemini, the visual interface should ideally support inputs from Video URLs and base64.

#### Image

```python
chat_response = client.chat.completions.create(
    model="llava",
    messages=[{
        "role": "user",
        "content": [
            {"type": "text", "text": "What’s in this image?"},
            {"type": "image_url", "image_url": {"url": f"data:image/png;base64,{image}"}},
        ],
    }],
)
```

#### Video
```python
chat_response = client.chat.completions.create(
    model="llava",
    messages=[{
        "role": "user",
        "content": [
            {"type": "text", "text": "What’s in this video?"},
            {"type": "video_url", "video_url": {"url": video_url}},
        ],
    }],
)
```

```python
chat_response = client.chat.completions.create(
    model="llava",
    messages=[{
        "role": "user",
        "content": [
            {"type": "text", "text": "What’s in this video?"},
            {"type": "video_url", "video_url": {"url": f"data:video/png;base64,{video_base64}"}},
        ],
    }],
)
```
The interface design mentioned above is a prototype.  I have implemented this interface in my pipeline.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by litianjian**
### Online video support for VLMs
vLLM already supports a large number of MultiModal Machine Learning visual models, some of which support image and video input，such as [Qwen2-VL](https://huggingface.co/Qwen/Qwen2-VL-2B-Instruct), [LLaVA-Onevision](https://huggingface.co/llava-hf/llava-onevision-qwen2-7b-ov-hf), etc. Referring to the implementation of image, this proposal adds support for video.

Refer to the visual interfaces of OpenAI [(vision](https://platform.openai.com/docs/guides/vision) and [video](https://cookbook.openai.com/examples/gpt_with_vision_for_video_understanding)) and Google Gemini, the visual interface should ideally support inputs from Video URLs and base64.

FIX #9842 

### Examples
```python
vllm serve llava-hf/llava-onevision-qwen2-7b-ov-hf --served-model-name hello --trust-remote-code
```

```python
try:
    from decord import VideoReader, cpu
except ImportError:
    pass

import base64
from io import BytesIO
from PIL import Image
import numpy as np
import requests

from openai import OpenAI
import time

openai_api_key = "123"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

def encode_video(video_path, max_frames=80, is_fps_sampling=True):
    if video_path.startswith("http") or video_path.startswith("https"):
        response = requests.get(video_path)
        if response.status_code == 200:
            video_path = BytesIO(response.content)
        else:
            print('failed to load the video')

        vr = VideoReader(video_path, ctx=cpu(0), num_threads=1)
        total_frame_num = len(vr)
        if is_fps_sampling:
            # FPS Sampling
            avg_fps = round(vr.get_avg_fps())
            frame_idx = [i for i in range(
                0, total_frame_num, avg_fps)]
            if len(frame_idx) > max_frames:
                uniform_sampled_frames = np.linspace(
                    0, total_frame_num - 1, max_frames, dtype=int
                )
                frame_idx = uniform_sampled_frames.tolist()
            print(frame_idx)
        else:
            # uniform sampling
            if total_frame_num > max_frames:
                uniform_sampled_frames = np.linspace(
                    0, total_frame_num - 1, max_frames, dtype=int
                )
                frame_idx = uniform_sampled_frames.tolist()
            else:
                frame_idx = [i for i in range(0, total_frame_num)]
            print(frame_idx)

        frames = vr.get_batch(frame_idx).asnumpy()
        print("actual frames", len(frames))
        
        base64_frames = []
        for frame in frames:
            img = Image.fromarray(frame)
            output_buffer = BytesIO()
            img.save(output_buffer, format="PNG")

            byte_data = output_buffer.getvalue()
            base64_str = base64.b64encode(byte_data).decode("utf-8")
            base64_frames.append(base64_str)
        return base64_frames


video_url = "https://raw.githubusercontent.com/EvolvingLMMs-Lab/sglang/dev/onevision_local/assets/jobs.mp4"
frames = encode_video(video_url, max_frames=32, is_fps_sampling=False)
images = []
images.extend(frames)
video_base64 = ",".join(images)

chat_response = client.chat.completions.create(
    model="hello",
    temperature=0,
    max_tokens=1024,
    messages=[{
        "role": "user",
        "content": [
            # NOTE: The prompt formatting with the video token `<video>` is not needed
            # since the prompt will be processed automatically by the API server.
            {"type": "text", "text": "Please describe the video comprehensively as much as possible."},
            {"type": "video_url", "video_url": {"url": video_url}},
        ],
    }],
)
print("Chat completion output:", chat_response.choices[0].message.content)


chat_response = client.chat.completions.create(
    model="hello", 
    temperature=0,
    max_tokens=1024,
    messages=[{
        "role": "user",
        "content": [
            # NOTE: The prompt formatting with the video token `<video>` is not needed
            # since the prompt will be processed automatically by the API server.
            {"type": "text", "text": "Please describe the video comprehensively as much as possible."},
            {"type": "video_url", "video_url": {"url": f"data:video/png;base64,{video_base64}"}},
        ],
    }],
)
print("Chat completion output:", chat_response.choices[0].message.content)

chat_response = client.chat.completions.create(
    model="hello",
    messages=[{
        "role": "user",
        "content": [
            # NOTE: The prompt formatting with the image token `<image>` is not needed
            # since the prompt will be processed automatically by the API server.
            {"type": "text", "text": "What’s in this image?"},
            {"type": "image_url", "image_url": {"url": f"data:image/png;base64,{images[0]}"}},
        ],
    }],
)
print("Chat completion output:", chat_response.choices[0].message.content)

```

### CODE REVIEW

File: vllm/multimodal/utils.py [hunk 1]
```diff
@@ -72,6 +72,81 @@ async def async_fetch_image(image_url: str,
...
+    video_path = BytesIO(b)
+    vr = decord.VideoReader(video_path, num_threads=1)
+    total_frame_num = len(vr)
+
+    if total_frame_num > num_frames:
+        uniform_sampled_frames = np.linspace(0,
+                                             total_frame_num - 1,
+                                             num_frames,
+                                             dtype=int)
+        frame_idx = uniform_sampled_frames.tolist()
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:97]: Can we use `range` here for simplicity?

File: vllm/multimodal/utils.py [hunk 2]
```diff
@@ -191,14 +276,15 @@ def rescale_image_size(image: Image.Image,
 def try_import_video_packages() -> Any:
     try:
         import cv2
+        import decord
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:279]: Can you add `decord` as an optional dependency for vLLM?

File: vllm/entrypoints/chat_utils.py [hunk 3]
```diff
@@ -52,6 +54,23 @@ class ChatCompletionContentPartAudioParam(TypedDict, total=False):
     """The type of the content part."""
 
 
+class VideoURL(TypedDict, total=False):
+    url: Required[str]
+    """
+    Either a URL of the video or a data URL with base64 encoded video data.
+
+    Learn more in the
+    [Vision guide](https://cookbook.openai.com/examples/gpt_with_vision_for_video_understanding).
```
Comment (DarkLight1337) [inline vllm/entrypoints/chat_utils.py:63]: I think this is a bit misleading as the OpenAI tutorial passes videos as a list of images, which is different from here where the entire video is being sent as a single item.

Comment (litianjian) [reply vllm/entrypoints/chat_utils.py:63]: OK, make sense. This reference is indeed ambiguous, so let's remove it.

File: vllm/multimodal/utils.py (hunk 2 above)
Comment (litianjian) [reply vllm/multimodal/utils.py:279]: Of course, how can we make this an optional dependency? like packages in audio?

Comment (DarkLight1337) [reply vllm/multimodal/utils.py:279]: Yep

File: vllm/multimodal/utils.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/multimodal/utils.py:97]: Hmm... now that I think of it, `np.linspace` is more readable. Let's keep it as is then>

**Review by DarkLight1337 — COMMENTED** [review]
The code looks good, but please add some tests to verify this.

**Comment by litianjian** [comment]
> The code looks good, but please add some tests to verify this.

OK , I have updated the tests.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for adding this!

**Comment by DarkLight1337** [comment]
It looks like the tests failed though, PTAL.

**Comment by litianjian** [comment]
> It looks like the tests failed though, PTAL.

The tests succeeded in my local machine.

**Comment by DarkLight1337** [comment]
It looks like `vllm[video]` isn't being installed for the test environment.

**Comment by xiaoajie738** [comment]
Excuse me, I have a question to ask, if we want to use vllm to call the qwen2-lv model, when we want to pass the video as input, if I extract the frame and crop it on the client to get the desired video frame sequence, call the chat interface of vllm by passing the picture sequence or call the chat interface by directly passing the video and the sampling frequency and the desired size implemented in this pr, will there be any difference in the results of these two calling methods? I understand that video passing video will bring more network bandwidth pressure and latency, thanks for the answer!

**Comment by DarkLight1337** [comment]
> Excuse me, I have a question to ask, if we want to use vllm to call the qwen2-lv model, when we want to pass the video as input, if I extract the frame and crop it on the client to get the desired video frame sequence, call the chat interface of vllm by passing the picture sequence or call the chat interface by directly passing the video and the sampling frequency and the desired size implemented in this pr, will there be any difference in the results of these two calling methods? I understand that video passing video will bring more network bandwidth pressure and latency, thanks for the answer!

The HF processor will be called regardless of whether you have done preprocessing beforehand. I am not sure whether HF processor is intelligent enough to return early if the image has already been cropped though.

**Comment by DarkLight1337** [comment]
> > > Excuse me, I have a question to ask, if we want to use vllm to call the qwen2-lv model, when we want to pass the video as input, if I extract the frame and crop it on the client to get the desired video frame sequence, call the chat interface of vllm by passing the picture sequence or call the chat interface by directly passing the video and the sampling frequency and the desired size implemented in this pr, will there be any difference in the results of these two calling methods? I understand that video passing video will bring more network bandwidth pressure and latency, thanks for the answer!
> > 
> > 
> > The HF processor will be called regardless of whether you have done preprocessing beforehand. I am not sure whether HF processor is intelligent enough to return early if the image has already been cropped though.
> 
> "Apologies if I wasn’t clear enough. I wanted to ask whether I can convert the video into individual frames and then call the chat interface by passing them as multiple images using the image_url field, rather than using the video_url field introduced in this PR. Is this approach feasible?"

Yes, Qwen2-VL supports both multi-image and video input.

**Comment by DarkLight1337** [comment]
Tests should pass now!

**Comment by litianjian** [comment]
> Tests should pass now!

 Thank you for your patience.

**Comment by hujh818** [comment]
@litianjian Thank you very much for your work. However, currently when video_url is passed in, we cannot control the logic of video frame extraction and image resizing. As a result, we cannot finely control the output result of the video. I am wondering if it is possible to specify a video_process function at the same time when the url is passed in, similar to encode_video when processing base64. After vllm downloads the video, use this function to process the video. The reason for hoping to adopt this method instead of inputting in base64 is that transmitting base64 video occupies too much bandwidth and is prone to network congestion.

**Comment by DarkLight1337** [comment]
For some models (e.g. Qwen2-VL), You can set `--mm-processor-kwargs` on startup to configure the HF processor class.

**Comment by wuxianyess** [comment]
> > > > Excuse me, I have a question to ask, if we want to use vllm to call the qwen2-lv model, when we want to pass the video as input, if I extract the frame and crop it on the client to get the desired video frame sequence, call the chat interface of vllm by passing the picture sequence or call the chat interface by directly passing the video and the sampling frequency and the desired size implemented in this pr, will there be any difference in the results of these two calling methods? I understand that video passing video will bring more network bandwidth pressure and latency, thanks for the answer!
> > > 
> > > 
> > > The HF processor will be called regardless of whether you have done preprocessing beforehand. I am not sure whether HF processor is intelligent enough to return early if the image has already been cropped though.
> > 
> > 
> > "Apologies if I wasn’t clear enough. I wanted to ask whether I can convert the video into individual frames and then call the chat interface by passing them as multiple images using the image_url field, rather than using the video_url field introduced in this PR. Is this approach feasible?"
> 
> Yes, Qwen2-VL supports both multi-image and video input.

If I use multiple images, how many images can  input at most, and how can I increase this maximum value?

**Comment by DarkLight1337** [comment]
Please refer to the `--limit-mm-per-prompt` [engine argument](https://docs.vllm.ai/en/latest/usage/engine_args.html).

**Pull request merged**